## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 458984`
Commit: `9e40e0571d65d20a4f94df0a4daef9ff0d24a64e`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>arc_unpacker</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>arc_unpacker</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>arc_unpacker</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>arc_unpacker</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>arc_unpacker</summary>
<pre>[717/736] Building CXX object CMakeFiles/run_tests.dir/tests/io/file_test.cc.o
[718/736] Building CXX object CMakeFiles/run_tests.dir/tests/io/file_byte_stream_test.cc.o
[719/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/common.cc.o
[720/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/file_support.cc.o
[721/736] Building CXX object CMakeFiles/run_tests.dir/tests/res/pixel_format_test.cc.o
[722/736] Building CXX object CMakeFiles/run_tests.dir/tests/flow/nested_decoding_error_test.cc.o
[723/736] Building CXX object CMakeFiles/run_tests.dir/tests/flow/unpacking_virtual_file_system_test.cc.o
[724/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/flow_support.cc.o
[725/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/decoder_support.cc.o
[726/736] Linking CXX executable arc_unpacker
[727/736] Building CXX object CMakeFiles/run_tests.dir/tests/flow/unpacking_recursion_test.cc.o
[728/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/image_support.cc.o
[729/736] Building CXX object CMakeFiles/run_tests.dir/tests/io/bit_stream_test.cc.o
[730/736] Building CXX object CMakeFiles/run_tests.dir/tests/res/image_test.cc.o
[731/736] Building CXX object CMakeFiles/run_tests.dir/tests/test_support/stream_test.cc.o
[732/736] Building CXX object CMakeFiles/run_tests.dir/tests/io/path_test.cc.o
[733/736] Building CXX object CMakeFiles/run_tests.dir/tests/docs_test.cc.o
[734/736] Building CXX object CMakeFiles/run_tests.dir/tests/types_test.cc.o
[735/736] Building CXX object CMakeFiles/run_tests.dir/tests/main.cc.o
ninja: build stopped: subcommand failed.</pre>
</details>
</details>

--- a/pkgs/by-name/ar/arc_unpacker/fix-test-float-variance.patch
+++ b/pkgs/by-name/ar/arc_unpacker/fix-test-float-variance.patch
@@ -1,0 +1,101 @@
+diff --git a/tests/test_support/audio_support.cc b/tests/test_support/audio_support.cc
+index c26022e9..5f50d035 100644
+--- a/tests/test_support/audio_support.cc
++++ b/tests/test_support/audio_support.cc
+@@ -16,6 +16,7 @@
+ // along with arc_unpacker. If not, see <http://www.gnu.org/licenses/>.
+ 
+ #include "test_support/audio_support.h"
++#include <cstdlib>
+ #include "algo/format.h"
+ #include "algo/range.h"
+ #include "dec/microsoft/wav_audio_decoder.h"
+@@ -44,13 +45,42 @@ res::Audio tests::get_test_audio()
+ void tests::compare_audio(
+     const res::Audio &actual, const res::Audio &expected)
+ {
++    // Allow for minor variances in audio samples within the specified
++    // threshold. Floating point calculations in some audio decoders yields
++    // slightly differing results across different CPU architectures.
++    //
++    // Affected tests:
++    //   * tests/dec/cri/hca_audio_decoder_test.cc
++    //   * tests/dec/entis/mio_audio_decoder_test.cc
++    static constexpr int tolerance = 1;
++
+     REQUIRE(actual.codec == expected.codec);
+     REQUIRE(actual.channel_count == expected.channel_count);
+     REQUIRE(actual.sample_rate == expected.sample_rate);
+     REQUIRE(actual.bits_per_sample == expected.bits_per_sample);
+     REQUIRE(actual.extra_codec_headers == expected.extra_codec_headers);
+     REQUIRE(actual.samples.size() == expected.samples.size());
++
++#ifdef __x86_64__
+     tests::compare_binary(actual.samples, expected.samples);
++#else
++    tests::compare_binary(
++      actual.samples,
++      expected.samples,
++      [](const bstr& bytes1, const bstr& bytes2) -> bool {
++        const auto samples1 = reinterpret_cast<const s16*>(bytes1.c_str());
++        const auto samples2 = reinterpret_cast<const s16*>(bytes2.c_str());
++        const auto n_samples = bytes1.size() / sizeof(*samples1);
++        // algo::range is a bit incomplete to work with std::all_of
++        for (const int i : algo::range(n_samples)) {
++          if (std::abs(samples1[i] - samples2[i]) > tolerance) {
++            return false;
++          }
++        }
++        return true;
++      }
++    );
++#endif
+ 
+     REQUIRE(actual.loops.size() == expected.loops.size());
+     for (const auto i : algo::range(expected.loops.size()))
+diff --git a/tests/test_support/common.cc b/tests/test_support/common.cc
+index b63ffa89..265ef7b1 100644
+--- a/tests/test_support/common.cc
++++ b/tests/test_support/common.cc
+@@ -40,3 +40,19 @@ void au::tests::compare_binary(const bstr &actual, const bstr &expected)
+     INFO(actual_dump << " != " << expected_dump);
+     REQUIRE(actual == expected);
+ }
++
++void au::tests::compare_binary(
++    const bstr &actual,
++    const bstr &expected,
++    std::function<bool(const bstr &, const bstr &)> compare)
++{
++    const auto max_size = 10000;
++    auto actual_dump = algo::hex(actual);
++    auto expected_dump = algo::hex(expected);
++    if (actual_dump.size() > max_size)
++        actual_dump = actual_dump.substr(0, max_size) + "(...)";
++    if (expected_dump.size() > max_size)
++        expected_dump = expected_dump.substr(0, max_size) + "(...)";
++    INFO(actual_dump << " != " << expected_dump);
++    REQUIRE(compare(actual, expected));
++}
+diff --git a/tests/test_support/common.h b/tests/test_support/common.h
+index 7d8be579..ecd5dcf8 100644
+--- a/tests/test_support/common.h
++++ b/tests/test_support/common.h
+@@ -17,6 +17,7 @@
+ 
+ #pragma once
+ 
++#include <functional>
+ #include "io/path.h"
+ #include "types.h"
+ 
+@@ -27,4 +28,9 @@ namespace tests {
+ 
+     void compare_binary(const bstr &actual, const bstr &expected);
+ 
++    void compare_binary(
++        const bstr &actual,
++        const bstr &expected,
++        std::function<bool(const bstr &, const bstr &)> compare);
++
+ } }

--- a/pkgs/by-name/ar/arc_unpacker/package.nix
+++ b/pkgs/by-name/ar/arc_unpacker/package.nix
@@ -27,23 +27,14 @@ stdenv.mkDerivation {
     hash = "sha256-STbdWH7Mr3gpOrZvujblYrIIKEWBHzy1/BaNuh4teI8=";
   };
 
-  patches = [
-    ./fix-test-float-variance.patch
-    (fetchpatch {
-      name = "failing_tests.patch";
-      url = "https://aur.archlinux.org/cgit/aur.git/plain/failing_tests.patch?h=arc_unpacker-git&id=bda1ad9f69e6802e703b2e6913d71a36d76cfef9";
-      hash = "sha256-bClACsf/+SktyLAPtt7EcSqprkw8JVIi1ZLpcJcv9IE=";
-    })
-    (fetchpatch {
-      name = "include_cstdint.patch";
-      url = "https://aur.archlinux.org/cgit/aur.git/plain/include_cstdint.patch?h=arc_unpacker-git&id=8c5c5121b23813c7650db19cb617b409d8fdcc9f";
-      hash = "sha256-3BQ1v7s9enUK/js7Jqrqo2RdSRvGVd7hMcY4iL51SiE=";
-    })
-  ];
+  patches = [ ./fix-test-float-variance.patch ];
 
   postPatch = ''
     cp ${catch2}/include/catch2/catch.hpp tests/test_support/catch.h
+
+    # missing includes
     sed '1i#include <limits>' -i src/dec/eagls/pak_archive_decoder.cc # gcc12
+    sed '1i#include <cstdint>' -i src/types.h # gcc13
     sed '1i#include <vector>' -i src/flow/cli_facade.h # gcc14
 
     # cmake-4 support
@@ -70,10 +61,18 @@ stdenv.mkDerivation {
   ];
 
   checkPhase = ''
+    # Specify test targets
+    # https://catch2-temp.readthedocs.io/en/latest/command-line.html#specifying-which-tests-to-run
+    checkTarget=('~CatSystem INT archives')
+
     runHook preCheck
 
+    local flagsArray=()
+    concatTo flagsArray checkFlags checkFlagsArray checkTarget
+
     pushd ..
-    ./build/run_tests
+    echoCmd 'check flags' "''${flagsArray[@]}"
+    ./build/run_tests "''${flagsArray[@]}"
     popd
 
     runHook postCheck

--- a/pkgs/by-name/ar/arc_unpacker/package.nix
+++ b/pkgs/by-name/ar/arc_unpacker/package.nix
@@ -27,6 +27,7 @@ stdenv.mkDerivation {
   };
 
   patches = [
+    ./fix-test-float-variance.patch
     (fetchpatch {
       name = "failing_tests.patch";
       url = "https://aur.archlinux.org/cgit/aur.git/plain/failing_tests.patch?h=arc_unpacker-git&id=bda1ad9f69e6802e703b2e6913d71a36d76cfef9";
@@ -88,8 +89,7 @@ stdenv.mkDerivation {
     runHook postInstall
   '';
 
-  # A few tests fail on aarch64-linux
-  doCheck = !(stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isAarch64);
+  doCheck = true;
 
   meta = with lib; {
     description = "Tool to extract files from visual novel archives";
@@ -98,8 +98,5 @@ stdenv.mkDerivation {
     maintainers = with maintainers; [ midchildan ];
     platforms = platforms.all;
     mainProgram = "arc_unpacker";
-
-    # unit test failures
-    broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
   };
 }

--- a/pkgs/by-name/ar/arc_unpacker/package.nix
+++ b/pkgs/by-name/ar/arc_unpacker/package.nix
@@ -5,6 +5,7 @@
   fetchpatch,
   cmake,
   makeWrapper,
+  ninja,
   boost,
   libpng,
   libiconv,
@@ -54,6 +55,7 @@ stdenv.mkDerivation {
   nativeBuildInputs = [
     cmake
     makeWrapper
+    ninja
     catch2
   ];
 

